bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/resource/CoreResourceWrapper.java

```diff
@@ -0,0 +1,80 @@
+/*~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ ~ Copyright 2021 Adobe
+ ~
+ ~ Licensed under the Apache License, Version 2.0 (the "License");
+ ~ you may not use this file except in compliance with the License.
+ ~ You may obtain a copy of the License at
+ ~
+ ~     http://www.apache.org/licenses/LICENSE-2.0
+ ~
+ ~ Unless required by applicable law or agreed to in writing, software
+ ~ distributed under the License is distributed on an "AS IS" BASIS,
+ ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ ~ See the License for the specific language governing permissions and
+ ~ limitations under the License.
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/
+package com.adobe.cq.wcm.core.components.internal.resource;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import org.apache.commons.lang3.StringUtils;
+import org.apache.sling.api.resource.Resource;
+import org.apache.sling.api.resource.ResourceResolver;
+import org.apache.sling.api.resource.ResourceWrapper;
+import org.apache.sling.api.resource.ValueMap;
+import org.apache.sling.api.wrappers.ValueMapDecorator;
+import org.apache.sling.models.annotations.Exporter;
+import org.jetbrains.annotations.NotNull;
+
+import com.adobe.cq.export.json.ExporterConstants;
+
+@Exporter(name = ExporterConstants.SLING_MODEL_EXPORTER_NAME , extensions = ExporterConstants.SLING_MODEL_EXTENSION)
+public class CoreResourceWrapper extends ResourceWrapper {
+
+    private ValueMap valueMap;
+    private String overriddenResourceType;
+
+    public CoreResourceWrapper(@NotNull Resource resource, @NotNull String overriddenResourceType, @NotNull Map<String, String> overriddenProperties) {
+        super(resource);
+        if (StringUtils.isEmpty(overriddenResourceType)) {
+            throw new IllegalArgumentException("The " + CoreResourceWrapper.class.getName() + " needs to override the resource type of " +
+                    "the wrapped resource, but the resourceType argument was null or empty.");
+        }
+        this.overriddenResourceType = overriddenResourceType;
+        valueMap = new ValueMapDecorator(new HashMap<>(resource.getValueMap()));
+        valueMap.put(ResourceResolver.PROPERTY_RESOURCE_TYPE, overriddenResourceType);
```

Comment:
This logic can be replaced with a org.apache.sling.api.wrappers.CompositeValueMap.

In general I would do this already in the component so the overridden and hidden properties collections get obsolete.

Reply:
`CompositeValueMap` doesn't do exactly what is needed here as it only merges the properties. There are use cases where we only want to remove properties, some where we only want to override properties and some where a combination of those is needed. I'd like to keep both `hiddenProperties` and `overriddenProperties` to have more flexibility.